// @matterlabs/hardhat-zksync-verify/dist/src/zksync-block-explorer/verification-status-response.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VerificationStatusResponse = void 0;
var VerificationStatusEnum;
(function (VerificationStatusEnum) {
    VerificationStatusEnum["successful"] = "successful";
    VerificationStatusEnum["failed"] = "failed";
    VerificationStatusEnum["queued"] = "queued";
    VerificationStatusEnum["inProgress"] = "in_progress";
})(VerificationStatusEnum || (VerificationStatusEnum = {}));
class VerificationStatusResponse {
    constructor(response) {
        this.status = response.data.status;
        this.error = response.data.error;
        this.compilationErrors = response.data.compilationErrors;
    }
    errorExists() {
        return !!(this.error || this.compilationErrors);
    }
    getError() {
        let errors = '';
        if (this.error) {
            errors += this.error;
        }
        if (this.compilationErrors) {
            errors += this.compilationErrors.join('\n');
        }
        return errors;
    }
    isPending() {
        return this.status === VerificationStatusEnum.inProgress;
    }
    isVerificationFailure() {
        return this.status === VerificationStatusEnum.failed;
    }
    isQueued() {
        return this.status === VerificationStatusEnum.queued;
    }
    isVerificationSuccess() {
        return this.status === VerificationStatusEnum.successful;
    }
}
exports.VerificationStatusResponse = VerificationStatusResponse;
